RavenDB-24565 AI Assistant endpoint for RavenDB

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24565

### Additional description


### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [ ] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

maciejaszyk committed:
RavenDB-24565: Expose a method to get the actual license in LicenseManager.

maciejaszyk committed:
RavenDB-24565 Initial work on endpoints for AI Assistant

Introduced:
- Consent endpoint
- Assist endpoint
- Quota endpoint

# Conflicts:
#	src/Raven.Server/Json/JsonDeserializationServer.cs

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/AiAssistantLimits.cs:
```diff
@@ -0,0 +1,9 @@
+
+public class AiAssistantLimits
+{
+    public float Quota { get; set; }
```
this is % ?

Lwiel replied:
Changed to `public int UsagePercent { get; set; }`

ppekrol replied:
UsagePercentage?

Lwiel replied:
Done

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/Processors/AiAssistantHandlerProcessorBase.cs:
```diff
@@ -0,0 +1,36 @@
+    protected AiAssistantHandlerProcessorBase([NotNull] RequestHandler requestHandler) : base(requestHandler)
+    {
+        // TODO Grisha
+        var licenseExists = ServerStore.LicenseManager.TryGetLicense(out _license);
```
ServerStore.LoadLicense

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/Processors/AiAssistantHandlerProcessorBase.cs:
```diff
@@ -0,0 +1,36 @@
+        if (licenseExists == false)
+            throw new InvalidOperationException($"AI Assistant is available only for licenses instances of RavenDB. Please register your license.");
+
+        _certificateThumbprint = RequestHandler.GetCurrentCertificate()?.Thumbprint ?? "unsecured node";
```
just `unsecured`

Lwiel replied:
Sending `null` if certificate isn't used

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/Processors/AiAssistantConsentProcessor.cs:
```diff
@@ -0,0 +1,44 @@
+                }
+            }
+
+            var responseString = await response.Content.ReadAsStringWithZstdSupportAsync().ConfigureAwait(false);
```
any status code support? predictable exceptions, exception schema, etc?

Lwiel replied:
We propagate status code returned in API response (if it's unavailable we also return relevant status code). We can add `ErrorMessage`/`ExceptionMessage` field to assistant response base class and fill it when exception occurs, as well as relevant value to `AiResponseStatus`. What do you think? @ppekrol @arekpalinski

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/Processors/AiAssistantAssistProcessor.cs:
```diff
@@ -0,0 +1,67 @@
+                }
+            }
+            
+            var responseString = await response.Content.ReadAsStringWithZstdSupportAsync().ConfigureAwait(false);
```
status code support, etc (check other comment)

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/AiAssistantHandler.cs:
```diff
@@ -0,0 +1,29 @@
+
+public class AiAssistantHandler : DatabaseRequestHandler
+{
+    [RavenAction("/databases/*/assistant/assist", "POST", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
also support sharding?

maciejaszyk replied:
added

maciejaszyk committed:
RavenDB-24565

- Takes license from cluster
- Refactor AsiAssistantLimit object since reset date is a first day of a month
- Added support for sharded database

Lwiel committed:
RavenDB-24565 Refactorization

Lwiel commented on src/Raven.Server/Json/JsonDeserializationServer.cs:
```diff
@@ -353,6 +356,8 @@ internal sealed class JsonDeserializationServer : JsonDeserializationBase
 
         internal static readonly Func<BlittableJsonReaderObject, StudioTasksHandler.AiModelsRequest> AiModelsRequest = GenerateJsonDeserializationRoutine<StudioTasksHandler.AiModelsRequest>();
         
+        internal static readonly Func<BlittableJsonReaderObject, RefineTextRequest> RefineTextGenAi = GenerateJsonDeserializationRoutine<RefineTextRequest>();
```
Just a note to rename this

Lwiel replied:
Done

arekpalinski commented on src/Raven.Server/Documents/AI/AiAssistant/AiOperationType.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Raven.Server.Documents.AI.AiAssistant;
+
+public enum AiOperationType
```
Too generic name I think. Maybe `AiAssistantOperationType`

Lwiel replied:
Done

arekpalinski commented on src/Raven.Server/Documents/AI/AiAssistant/Responses/AiResponseStatus.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace Raven.Server.Documents.AI.AiAssistant.Responses;
+
+public enum AiResponseStatus
```
`AiAssistResponseStatus` maybe? Please make sure that we use consistent naming

Lwiel replied:
`AiAssistantResponseStatus`, done

Lwiel committed:
RavenDB-24565 Applied PR comments

Lwiel committed:
RavenDB-24565 Added property do AiAssistResponseBase

Lwiel committed:
RavenDB-24565 Added GenAi prompt refine handling

Lwiel committed:
RavenDB-24565 Post-rebase fix

ppekrol commented on src/Raven.Server/Documents/AI/AiAssistant/Handlers/Processors/AiAssistantHandlerProcessorBase.cs:
```diff
@@ -0,0 +1,35 @@
+    protected void FulfillRequestMetadata<TRequest>(TRequest request) where TRequest : AiAssistantRequestAuthentication
+    {
+        PortableExceptions.ThrowIf<InvalidDataException>(request.License is not null, "License should be set by the server.");
+        PortableExceptions.ThrowIf<InvalidDataException>(request.CertificateThumbprint is not null, "Certificate thumbprint should be set by the server.");
```
this can be null as well

Lwiel replied:
It checks if license and thumbprint weren't already set in request sent to Raven server. I don't think it makes sense to check it - we should just always override it instead

ppekrol commented on src/Raven.Server/Json/JsonDeserializationServer.cs:
```diff
@@ -84,6 +84,12 @@
 using FacetSetup = Raven.Client.Documents.Queries.Facets.FacetSetup;
 using MigrationConfiguration = Raven.Server.Smuggler.Migration.MigrationConfiguration;
 using StudioConfiguration = Raven.Client.Documents.Operations.Configuration.StudioConfiguration;
+using Raven.Server.Documents.ETL.Providers.AI.Embeddings.Test;
```
warnings here

Lwiel replied:
Cleared

ppekrol reviewed (approved):
small comments

arekpalinski commented on src/Raven.Server/Documents/AI/AiAssistant/Requests/RefineGenAiPromptRequest.cs:
```diff
@@ -0,0 +1,9 @@
+    public string SourceCollectionName { get; set; }
+    public string ContextGenerationScript { get; set; }
+    public string Prompt { get; set; }
+    public string ResponseFormat { get; set; }
```
The name might be a bit misleading. It sounds that you can choose the format of the response that will be required as the result of your request. While it applies to `Sample response object / JSON schema` I believe. Is that correct?

Lwiel replied:
Renamed to `OutputStructure`

Lwiel committed:
RavenDB-24565 Applied PR comments

Lwiel committed:
RavenDB-24565 Changed AiAssistantHandler to server-wide

kalczur commented:
rebased

kalczur committed:
RavenDB-24372 Studio initial work

kalczur commented:
I'm working on the studio part. I will let you know when it will be completed

kalczur committed:
RavenDB-24372 Move logic to redux and add story

kalczur committed:
RavenDB-24372 Add usage and consent to about page

kalczur committed:
RavenDB-24372 Styling

kalczur committed:
RavenDB-24372 Button styling

kalczur committed:
RavenDB-24372 Improve text

kalczur committed:
RavenDB-24372 Fix unnecessary consent checks

kalczur committed:
RavenDB-24372 Update consentStatus after giveConsent success

kalczur committed:
RavenDB-24372 Add spinner to give consent

kalczur commented:
The stuido part is ready. It only requires updating EULA link

M4xymm commented on src/Raven.Studio/typescript/components/common/aiAssistant/AiAssistantWindow.stories.tsx:
```diff
@@ -0,0 +1,70 @@
+        <div>
+            <div className="position-relative" style={{ width: "500px", height: "250px" }}>
+                <AiAssistantWindow
+                    closeWindow={() => {}}
```
empty function

kalczur replied:
It's in storybook (we don't want to close it here)

<img width="903" height="408" alt="image" src="https://github.com/user-attachments/assets/493fd934-214a-412e-9b8d-4846c41c08c9" />

M4xymm commented on src/Raven.Studio/typescript/commands/aiAssistant/assistAiAssistantCommand.ts:
```diff
@@ -0,0 +1,39 @@
+      } & Omit<
+          Raven.Server.Documents.AI.AiAssistant.Requests.RefineGenAiPromptRequest,
+          "OperationType" | "CertificateThumbprint" | "License"
+      >);
```
```suggestion
type AiRequest<
    T extends Raven.Server.Documents.AI.AiAssistant.AiAssistantOperationType,
    R
> = {
    OperationType: Extract<Raven.Server.Documents.AI.AiAssistant.AiAssistantOperationType, T>;
} & Omit<R, "OperationType" | "CertificateThumbprint" | "License">;

export type AssistAiAssistantRequestDto =
    | AiRequest<"RefineText", Raven.Server.Documents.AI.AiAssistant.Requests.RefineTextRequest>
    | AiRequest<"RefineGenAiPrompt", Raven.Server.Documents.AI.AiAssistant.Requests.RefineGenAiPromptRequest>;

```

M4xymm replied:
I think this will be more readable

kalczur replied:
much cleaner, thanks!

M4xymm commented on src/Raven.Studio/typescript/test/mocks/services/MockAiAssistantService.ts:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    withCheckUsage(dto?: MockedValue<CheckUsageAiAssistantResultDto>) {
+        return this.mockResolvedValue(this.mocks.checkUsage, dto, { Status: "Success", UsagePercentage: 35 });
```
I think it would be worth adding the `AiAssistantStubs.ts` file.

kalczur committed:
RavenDB-24565 Cleanup

kalczur committed:
RavenDB-24372 Add ai assistant args to about page story

kalczur commented:
Rebased

kalczur committed:
RavenDB-24565 Fix lint
